Minimize crypto operations to return an entry

kamidon opened the pull request:
Before this commit, multiple cryptographic operations were required to
return a single result. This doesn't cause any significant issue when
these operations obtain their key material from gpg-agent and no user
interaction beyond possibly an initial passphrase being entered is
required. However, there are some usage scenarios where it becomes
inconvenient. The example that motivated this commit is storing the key
that protects the password store on a PGP smartcard that requires a
physical action like a touch of a sensor to release the result of a
cryptographic operation. In these cases multiple and sometime a
variable number of physical interactions with the smartcard are
require to obtain a single entry.

This commit includes two changes that result in the common cases
requiring only a single physical interaction with the smartcard to
retrieve an entry:

1. `auth-source-pass-get' is split into two pieces, a backend called
   `auth-source-pass--get-attr' that returns the value of a key from
   already parsed entry data and which is then used by the original
   function.  `auth-source-pass--build-result' is then modified to
   parse the entry once and use `auth-source-pass-get-attr'.

2. A new defcustom, `auth-source-pass-validate-during-search' is
   introduced to allow the user to disable the validation of an entry
   during the search to locate applicable entries. This prevents
   cryptographic operations from being required during the search
   process. It does have a downside however in that it changes the
   behavior of the search and could cause users that are finding
   entries successfully to start having errors if they have multiple
   applicable entries in their password store, some of which don't
   parse successfully. For this reason the custom is set to retain the
   current behavior for default.

I've been testing a version with these changes applied and it has been
working well for me.

kamidon commented:
I think the most recent Travis failure was due to an environmental factor and nothing specific to the updated PR, which I hope should pass if it can be run successfully.

kamidon commented:
Thanks for the suggestions. I'll adopt the specific comments soon, though the same comment about my near-term schedule as I made in PR80 applies. 

I went the defcustom route primarily because solutions that still validate during search seemed like they would require fairly extensive changes to the call path for returning results.  I think the most straightforward solution would be to parse entries during the search and return the entry data up the call stack. Does that seem like a good approach?

DamienCassou commented:
>  I think the most straightforward solution would be to parse entries during the search and return the entry data up the call stack

I'm not sure I understand. My idea was to rewrite the algorithm so that:

1. a function returns the list of entries in the password store that match the request by only looking at the entry names (including directories). This list is sorted by most specific to least specific
2. another function iterates the list by parsing each entry and returns the first element whose content matches the query

What do you think?

DamienCassou commented:
>  I think the most straightforward solution would be to parse entries during the search and return the entry data up the call stack

I'm not sure I understand. My idea was to rewrite the algorithm so that:

1. a function returns the list of entries in the password store that match the request by only looking at the entry names (including directories). This list is sorted by most specific to least specific
2. another function iterates the list by parsing each entry and returns the first element whose content matches the query

What do you think?

kamidon commented:
The reason I was suggesting actually parsing the entries during the search is that I believe the way the search works right now it returns as soon as a matching entry that parses is found, short-circuiting the rest of the search. However, since there aren't many options to be searched anyway, it shoudn't hurt to check all of them for existence. I'll give that a try as soon as I can get a few hours to work on it.

kamidon commented:
I'm thinking of abandoning this PR and submitting two PRs in its place, one that just does the refactor of `auth-source-pass-get` & the its' use in `auth-source-pass--build-result` followed by a second PR that implements the suggested deferral of entry parsing until all possible matching entries have been found. I think that second PR will be fairly large so separating it from simple refactor (and the fix to the search behavior in PR 80) is the way to go. Sound good to you?

DamienCassou commented:
Please go ahead with 2 PRs. I think this is better.

DamienCassou commented:
@kamidon There is some discussion in another issue regarding the number of decryptions: https://github.com/DamienCassou/auth-password-store/issues/83#issuecomment-474720044.

kamidon commented:
Thanks for the heads up @DamienCassou. The FSF is currently reviewing my company's employer disclaimer so hopefully we're close to being in the clear to merge an updated version of these PRs that addresses all review feedback. I plan to spend some time on that as soon as everything is wrapped up with the FSF.

kamidon commented:
I hope to get to replacing this PR with the two we were discussing earlier tomorrow.

DamienCassou commented:
Can we close this PR?

kamidon commented:
Yeah, I think we can continue the discussion about how to address the needs that led to this PR in PR #89.
